File: 2023/08.07/정호조/컵라면.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class 컵라면_1781 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        PriorityQueue<Homework> homework = new PriorityQueue<>((o1, o2) ->{
           if(o1.deadLine == o2.deadLine){
               return o2.ramen - o1.ramen;
           } else {
               return o1.deadLine - o2.deadLine;
           }
        });
        PriorityQueue<Integer> result = new PriorityQueue<>();


        for(int i=0; i<N; i++){
            st = new StringTokenizer(br.readLine());
            int deadLine = Integer.parseInt(st.nextToken());
            int ramen = Integer.parseInt(st.nextToken());

            homework.add(new Homework(deadLine, ramen));
        }

        int sum =0;
        for(int i=0; i<N; i++){
            Homework hwTemp = homework.poll();
            result.add(hwTemp.ramen);

            if(hwTemp.deadLine < result.size()){
//                System.out.println("poll " + result.peek() + " deadline : "+hwTemp.deadLine + " size : "+result.size());
                result.poll();
            }
        }

        while (!result.isEmpty()){
            sum += result.poll();
        }

        System.out.println(sum);
    }

    static class Homework {
        int deadLine, ramen;

        public Homework(int deadLine, int ramen){
            this.deadLine = deadLine;
            this.ramen = ramen;
        }
    }
}

File: 2023/08.07/조수연/컵라면.kt
import java.io.*
import java.util.*

lateinit var st: StringTokenizer

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val n = readLine().toInt()
    val pq = PriorityQueue(compareBy<Pair<Int, Int>> { it.first }.thenByDescending { it.second })

    repeat(n) {
        st = StringTokenizer(readLine())
        pq.add(Pair(st.nextToken().toInt(), st.nextToken().toInt()))
    }

    val resultPq = PriorityQueue<Int>() // 라면 갯수가 적은 순으로 정렬
    for (i in 1..n) {
        val problem = pq.poll()
        resultPq.add(problem.second)

        if (problem.first < resultPq.size){ // 해당 데드라인에 문제가 여러개일 경우 갯수가 가장 높은 라면만 남기고 poll
            resultPq.poll()
        }
    }

    var answer = 0

    while (resultPq.isNotEmpty()){
        answer += resultPq.poll()
    }

    bw.write("$answer")

    bw.flush()
    bw.close()
}